syntax = "proto3";

package posts;

message Post {
    string id = 1;
    string title = 2;
    string description = 3;
    string creator_id = 4;
    string created_at = 5;
    string updated_at = 6;
    bool is_private = 7;
    repeated string tags = 8;
}

message CreatePostRequest {
    string title = 1;
    string description = 2;
    string creator_id = 3;
    bool is_private = 4;
    repeated string tags = 5;
}

message CreatePostResponse {
    Post post = 1;
}

message GetPostRequest {
    string id = 1;
}

message GetPostResponse {
    Post post = 1;
}

message UpdatePostRequest {
    string id = 1;
    string title = 2;
    string description = 3;
    bool is_private = 4;
    repeated string tags = 5;
}

message UpdatePostResponse {
    Post post = 1;
}

message DeletePostRequest {
    string id = 1;
}

message DeletePostResponse {
    string message = 1;
}

message ListPostsRequest {
    int32 page = 1;
    int32 page_size = 2;
}

message ListPostsResponse {
    repeated Post posts = 1;
}

message ViewRequest {
    string post_id = 1;
    string user_id = 2;
}
message ViewResponse {string message = 1;}

message LikeRequest {
    string post_id = 1;
    string user_id = 2;
}
message LikeResponse {string message = 1;}

message Comment {
    string id = 1;
    string post_id = 2;
    string user_id = 3;
    string content = 4;
    string created_at = 5;
}
message CreateCommentRequest {
    string post_id = 1;
    string user_id = 2;
    string content = 3;
}
message CreateCommentResponse {
    Comment comment = 1;
}

message ListCommentsRequest {
    string post_id = 1;
    int32 page = 2;
    int32 page_size = 3;
}
message ListCommentsResponse {
    repeated Comment comments = 1;
}


service PostService {
    rpc CreatePost (CreatePostRequest) returns (CreatePostResponse);
    rpc GetPost (GetPostRequest) returns (GetPostResponse);
    rpc UpdatePost (UpdatePostRequest) returns (UpdatePostResponse);
    rpc DeletePost (DeletePostRequest) returns (DeletePostResponse);
    rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);
    rpc ViewPost (ViewRequest) returns (ViewResponse);
    rpc LikePost (LikeRequest) returns (LikeResponse);
    rpc CreateComment (CreateCommentRequest) returns (CreateCommentResponse);
    rpc ListComments (ListCommentsRequest) returns (ListCommentsResponse);
}